From: Alfonso Sanchez-Beato <alfonso.sanchez-beato@canonical.com>
Date: Fri, 20 Feb 2015 14:08:40 +0100
Subject: [PATCH] qtdemux: Set maximum sample input size in caps

Read the maximum sample size from stsz atom and set max-input-size
capability with that value. This helps decoders that re-use input
buffers to select the right size for them. The size is determined
using the sample size field in stsz if set, or searching the maximum
sample size in the sample size table field otherwise.

https://bugzilla.gnome.org/show_bug.cgi?id=744857

Origin: vendor
Forwarded: yes

---
 gst/isomp4/qtdemux.c | 31 +++++++++++++++++++++++++++++++
 1 file changed, 31 insertions(+)

diff --git a/gst/isomp4/qtdemux.c b/gst/isomp4/qtdemux.c
index b766a38..e624888 100644
--- a/gst/isomp4/qtdemux.c
+++ b/gst/isomp4/qtdemux.c
@@ -6516,6 +6516,8 @@ flow_failed:
 static gboolean
 qtdemux_stbl_init (GstQTDemux * qtdemux, QtDemuxStream * stream, GNode * stbl)
 {
+  gint max_size = 0;
+
   stream->stbl_index = -1;      /* no samples have yet been parsed */
   stream->sample_index = -1;
 
@@ -6602,6 +6604,35 @@ qtdemux_stbl_init (GstQTDemux * qtdemux, QtDemuxStream * stream, GNode * stbl)
   if (!stream->n_samples)
     goto no_samples;
 
+  if (stream->sample_size == 0) {
+    GstByteReader max_sz_reader;
+    int i;
+
+    if (gst_byte_reader_get_remaining (&stream->stsz) != 4 * stream->n_samples)
+      goto corrupt_file;
+
+    gst_byte_reader_init (&max_sz_reader, stream->stsz.data + 1 + 3 + 4 + 4,
+        4 * stream->n_samples);
+
+    for (i = 0; i < stream->n_samples; ++i) {
+      guint32 cur;
+      cur = gst_byte_reader_get_uint32_be_unchecked (&max_sz_reader);
+      if (cur > max_size)
+        max_size = cur;
+    }
+  } else {
+    max_size = stream->sample_size;
+  }
+
+  /* Assume that a given buffer only contains at most 10 chunks,
+   * each chunk originally prefixed with a 2 byte length will
+   * have a 4 byte header (0x00 0x00 0x00 0x01) after conversion,
+   * and thus will grow by 2 bytes per chunk.
+   */
+  if (max_size > 0)
+    gst_caps_set_simple (stream->caps,
+        "max-input-size", G_TYPE_INT, max_size + 10 * 2, NULL);
+
   /* sample-to-chunk atom */
   if (!qtdemux_tree_get_child_by_type_full (stbl, FOURCC_stsc, &stream->stsc))
     goto corrupt_file;
-- 
2.1.0

